File: src/chaincfg.ts
import { init, g, unwrap } from './init';
import type { Network } from './types';

export interface ChainParams {
  name: string;
  defaultPort: string;
  bech32HRPSegwit: string;
  pubKeyHashAddrID: number;
  scriptHashAddrID: number;
  privateKeyID: number;
  hdPrivateKeyID: string;
  hdPublicKeyID: string;
  hdCoinType: number;
  coinbaseMaturity: number;
  subsidyReductionInterval: number;
}

/** Bitcoin network configuration parameters. */
export const chaincfg = {
  /** Get the full parameter set for a named network.
   *  Calls Go: chaincfg.MainNetParams / TestNet3Params / etc. from btcd/chaincfg. */
  async getParams(network: Network): Promise<ChainParams> {
    await init();
    return unwrap<ChainParams>(g().chaincfg.getParams(network));
  },
  /** Check if a byte is a known P2PKH address prefix.
   *  Calls Go: chaincfg.IsPubKeyHashAddrID() from btcd/chaincfg. */
  async isPubKeyHashAddrID(id: number): Promise<boolean> {
    await init();
    return unwrap<boolean>(g().chaincfg.isPubKeyHashAddrID(id));
  },
  /** Check if a byte is a known P2SH address prefix.
   *  Calls Go: chaincfg.IsScriptHashAddrID() from btcd/chaincfg. */
  async isScriptHashAddrID(id: number): Promise<boolean> {
    await init();
    return unwrap<boolean>(g().chaincfg.isScriptHashAddrID(id));
  },
  /** Check if a string is a known bech32 segwit HRP prefix.
   *  Calls Go: chaincfg.IsBech32SegwitPrefix() from btcd/chaincfg. */
  async isBech32SegwitPrefix(prefix: string): Promise<boolean> {
    await init();
    return unwrap<boolean>(g().chaincfg.isBech32SegwitPrefix(prefix));
  },
  /** Convert an HD private key version to its public counterpart.
   *  Calls Go: chaincfg.HDPrivateKeyToPublicKeyID() from btcd/chaincfg. */
  async hdPrivateKeyToPublicKeyID(hexPrivateKeyID: string): Promise<Uint8Array> {
    await init();
    return unwrap<Uint8Array>(g().chaincfg.hdPrivateKeyToPublicKeyID(hexPrivateKeyID));
  },
};

File: wasm-wrapper/chaincfg.go
//go:build js && wasm

package main

import (
	"encoding/hex"
	"syscall/js"

	"github.com/btcsuite/btcd/chaincfg"
)

func chaincfgGetParams(_ js.Value, args []js.Value) any {
	if e := checkArgs(args, 1, "network"); e != nil {
		return e
	}
	params, e := getNetwork(args[0].String())
	if e != nil {
		return e
	}
	return okResult(map[string]any{
		"name":                     params.Name,
		"defaultPort":              params.DefaultPort,
		"bech32HRPSegwit":          params.Bech32HRPSegwit,
		"pubKeyHashAddrID":         int(params.PubKeyHashAddrID),
		"scriptHashAddrID":         int(params.ScriptHashAddrID),
		"privateKeyID":             int(params.PrivateKeyID),
		"hdPrivateKeyID":           hex.EncodeToString(params.HDPrivateKeyID[:]),
		"hdPublicKeyID":            hex.EncodeToString(params.HDPublicKeyID[:]),
		"hdCoinType":               int(params.HDCoinType),
		"coinbaseMaturity":         int(params.CoinbaseMaturity),
		"subsidyReductionInterval": int(params.SubsidyReductionInterval),
	})
}

func chaincfgIsPubKeyHashAddrID(_ js.Value, args []js.Value) any {
	if e := checkArgs(args, 1, "id"); e != nil {
		return e
	}
	return okResult(chaincfg.IsPubKeyHashAddrID(byte(args[0].Int())))
}

func chaincfgIsScriptHashAddrID(_ js.Value, args []js.Value) any {
	if e := checkArgs(args, 1, "id"); e != nil {
		return e
	}
	return okResult(chaincfg.IsScriptHashAddrID(byte(args[0].Int())))
}

func chaincfgIsBech32SegwitPrefix(_ js.Value, args []js.Value) any {
	if e := checkArgs(args, 1, "prefix"); e != nil {
		return e
	}
	return okResult(chaincfg.IsBech32SegwitPrefix(args[0].String()))
}

func chaincfgHDPrivateKeyToPublicKeyID(_ js.Value, args []js.Value) any {
	if e := checkArgs(args, 1, "hexPrivateKeyID"); e != nil {
		return e
	}
	b, e := hexDecode(args[0].String())
	if e != nil {
		return e
	}
	pubID, err := chaincfg.HDPrivateKeyToPublicKeyID(b)
	if err != nil {
		return errfResult("hdPrivateKeyToPublicKeyID: %s", err)
	}
	return okResult(hex.EncodeToString(pubID))
}
